### PR TITLE
Sentinel 2

### PR DESCRIPTION
**Pull request by Marie59**
OK here we go again ;) I hope it is newest version with all the latest versions with all your fixes @bgruening taking into account but I am not sure (I pulled this version not long ago but maybe there were some changes made after I don't remember)

### CODE REVIEW

**Comment by Marie59** [comment]
I have a brand new issue that just appeared from nowhere in my R script. I'm on it !

**Comment by yvanlebras** [comment]
seems that the " srs_preprocess_s2 (Test #1)" error is due to a memory issue of the CI container. For "srs_metadata (Test #1)" duno. For "srs_diversity_maps", `Error in biodivMapR::map_spectral_species(Input_Image_File = input_image_file,  :`

**Comment by Marie59** [comment]
> seems that the " srs_preprocess_s2 (Test #1)" error is due to a memory issue of the CI container. For "srs_metadata (Test #1)" duno. For "srs_diversity_maps", `Error in biodivMapR::map_spectral_species(Input_Image_File = input_image_file, :`

For map_spectral_species I am trying things but for now nothing is working. I think it comes from an upgrade of biodivmapr.

**Comment by Marie59** [comment]
Yes I have this error too on my machine but it is from the package biodivmapr and it does not preventthe tool from working. It is indeed running as it should as these messages are saying.

**Comment by bgruening** [comment]
So the tool is working but it writes to stderr? Or exists with an error code?

File: tools/srs_tools/val_metadata.xml [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+        <requirement type="package" version="0.7">r-geometa</requirement>
+        <requirement type="package" version="1.4.3">r-httr</requirement>
+        <requirement type="package" version="3.99_0.9">r-xml</requirement>
+        <requirement type="package" version="1.3.3">r-xml2</requirement>
+    </requirements>
+    <command detect_errors="exit_code"><![CDATA[
+        Rscript
+            '$__tool_directory__/val_metadata.r'
+            '$input'
+            '$output' 1> Metadata_validation.txt
```
Comment (bgruening) [inline tools/srs_tools/val_metadata.xml:25]: @Marie59 is it correct that you only want to use the output from stdout and NOT what is written to $output?

You are picking with `from_work_dir="Metadata_validation.txt"` so I think $output is ignored.

**Comment by bgruening** [comment]
Locally I get this here:

```
Error in .rasterObjectFromFile(x, band = band, objecttype = "RasterLayer",  : 
  Cannot create a RasterLayer object from this file.
Calls: <Anonymous> ... <Anonymous> -> <Anonymous> -> .local -> .rasterObjectFromFile
Execution halted
```

This looks like an R code problem? Can I help here?

**Comment by Marie59** [comment]
> Locally I get this here:
> 
> ```
> Error in .rasterObjectFromFile(x, band = band, objecttype = "RasterLayer",  : 
>   Cannot create a RasterLayer object from this file.
> Calls: <Anonymous> ... <Anonymous> -> <Anonymous> -> .local -> .rasterObjectFromFile
> Execution halted
> ```
> 
> This looks like an R code problem? Can I help here?

For map_diversity ? weird I fixed that normally. I'm gonna check

**Comment by Marie59** [comment]
Oh I think I know it's because on my machine locally I implemented 2 datatypes the bil one and the hdr for the test files

**Comment by bgruening** [comment]
Where are they, should we get them upstream? Is any help needed? I would like to help you get this thing in before Christmas :)

**Comment by Marie59** [comment]
> Where are they, should we get them upstream? Is any help needed? I would like to help you get this thing in before Christmas :)

So sorry I did not see this last message until now. Yes I think we have to put them upstream. I don't really know how to do it. I added them on my local galaxy. I can check tonight on how to add those datatypes

**Comment by bgruening** [comment]
If you get stuck let me know or sent me the changes

**Comment by Marie59** [comment]
Ok I'm a bit lost. I found the lines where I added those datatypes and followed the guidelines to add datatypes but I don't see where to go to do some kind of PR to add it on galaxy. Sorry

**Comment by yvanlebras** [comment]
It seems to me we here have the same errors than me locally :

**Comment by bgruening** [comment]
@yvanlebras do you use `planemo test --biocontainers` to test the tools?

**Comment by yvanlebras** [comment]
> 

nope, I am on a personal Galaxy installation, not using planemo

**Comment by yvanlebras** [comment]
It seems to be better now! 
- srs_metadata is ok
- srs_spectral_indices failed but seems to be linked to a difference in resulting files, not a running issue
- srs_process_data is ok

Duno why others are not available, apparently due to a cancellation of the GitHub actions

**Comment by Marie59** [comment]
I just fixed spectral_indices and it seems to be ok now !

**Comment by yvanlebras** [comment]
OK for val_metadata ! 

image

SO, I have to try biodiv indices locally, as it seems maybe this is not working on github actions due to too much resources need..

**Comment by yvanlebras** [comment]
For biodiv_indice, trying loccaly to create the same conda env, it works, I mean, I can execute Rscript command after activating the conda env.... But with the Galaxy tool, we still have such "Rscript: command not found".

For info, the command I used to create a local similar conda env: `conda create --name biodivindice r-base=4.2.2 r-raster=3.5_21 r-rgdal=1.5_32 r-sp=1.5_1 r-rasterdiv=0.3.1 r-ggplot2=3.4.0 r-terra=1.5_21 -c conda-forge`

**Comment by yvanlebras** [comment]
Ok, I tried all tools locally, and this is ok with these xml wrappers versions! THANK YOU Marie ! Hope it will be ok once on eu!

**Review by yvanlebras — APPROVED** [review]
This is ok for me now! Amazing wonder Marie !!! THANK You Marie, Björn, for all your huge efforts on it! Now we have to communicate ++ on it! Yes, we can analyze satellite remote sensing data on Galaxy now!!!!!!!!!

**Comment by yvanlebras** [comment]
is it ok for you Björn to merge?

**Comment by yvanlebras** [comment]
oh, important info considering `alpha_beta.xml` tool. In the test section, the parameter is set to "all" to compute all indices (alpha+beta+functional) BUT it seems to ask for large amount of memory or something like that and it fail. Locally, I asked for computing alpha indices in one hand, then beta, then functional, and it was ok one by one for each!

**Comment by yvanlebras** [comment]
So, looking at job resources management, it seems important to tune these wrappers to limit memory / cpu usage https://planemo.readthedocs.io/en/latest/writing_advanced.html?highlight=Galaxy_slots#developing-for-clusters-galaxy-slots-galaxy-memory-mb-and-galaxy-memory-mb-per-slot

**Comment by bgruening** [comment]
CPU is the important one I think. Not sure where to do that in the R code.

**Comment by bgruening** [comment]
down to 1 failing tool, and this is due to too many cores I assume

File: tools/srs_tools/functions.r [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+# too small window_size results in low number of pixels per spatial unit, hence limited range of variation of diversity in the image
+window_size <- 10
+
+# PCA FILTERING: Set to TRUE if you want second filtering based on PCA outliers to be processed. Slower
+filterpca <- TRUE
+
+################################################################################
+##                    DEFINE PARAMETERS FOR METHOD                            ##
+################################################################################
+nbcpu <- 4
```
Comment (bgruening) [inline tools/srs_tools/functions.r:36]: This should be passed from the Galaxy tool XML using GALAXY_SLOTS

**Comment by yvanlebras** [comment]
I remove the input test-data of preprocess to try uploading a new version lighter.. but I don't have the push access to upload it  @Marie59  ;)

**Comment by bgruening** [comment]
Sent it to me ... ;)

**Comment by bgruening** [comment]
Let's merge and push manually. The last failing test is due to much more memory consumption. it would be nice to have smaller test data set, that in turn also needs less memory.

**Pull request merged**